csrc/includes/quantization_utils.h

```diff
@@ -145,112 +146,85 @@ class Params<Type::Asymmetric, numBits> {
+    return __habs2(val);
+}
+
+DS_D_INLINE float to_max_float(const __half& val) {
+    return __half2float(val);
+}
+
+DS_D_INLINE float to_min_float(const __half& val) {
+    return __half2float(val);
+}
+
+DS_D_INLINE float to_max_float(const __half2& val) {
+    const float2 partial_max = conversion::to<float2>(val);
+    return reduce::element<rop::Max>(partial_max.x, partial_max.y);
+}
 
-template <>
-class GroupStats<Type::Symmetric> {
+DS_D_INLINE float to_min_float(const __half2& val) {
+    const float2 partial_min = conversion::to<float2>(val);
+    return reduce::element<rop::Min>(partial_min.x, partial_min.y);
+}
+
+// GroupStats class template
+template <Type qType, typename DataType = __half2>
+class GroupStats;
+
+// Symmetric Quantization
+template <typename DataType>
+class GroupStats<Type::Symmetric, DataType> {
 public:
-    // Symmetric quantization only tracks the maximum absolute value
-    __half2 cur_max;
-    float max;
+    DataType cur_max;
 
-    /*
-    Technically, this would give bad results if there
-    are 0 values to process since the reduction would
-    give -inf instead of 0. We do not consider this
-    to be a reasonable edge case.
-    */
-    DS_D_INLINE GroupStats() { cur_max = reduce::init<rop::Max, __half2>(); }
+    DS_D_INLINE GroupStats() { cur_max = reduce::init<rop::Max, DataType>(); }
 
-    /*
-    Updated the running absmax used to calculate params.
-    Function Arguments :
-        val : The __half2 value to update the running min and max with.
-    */
-    DS_D_INLINE void update(__half2 val)
-    {
-        cur_max = reduce::element<rop::Max>(cur_max, __habs2(val));
+    DS_D_INLINE void update(DataType val) {
+        cur_max = reduce::element<rop::Max>(cur_max, h_abs(val));
     }
 
-    /*
-    Function to return calculated quantization params.
-    Template Arguments :
-        numBits -   Number of bits in quantized element.    int : 8 or 4
-    Function Arguments :
-        tb      -   Threadblock object. cg::thread_block
-        warp    -   Warp object.        cg::thread_block_tile<hw_warp_size>
-    */
     template <int numBits, int threads_per_group>
     DS_D_INLINE Params<Type::Symmetric, numBits> get_params(
         cg::thread_block& tb,
-        cg::thread_block_tile<hw_warp_size>& warp)
-    {
-        const float2 partial_max = conversion::to<float2>(cur_max);
-        float max = reduce::element<rop::Max>(partial_max.x, partial_max.y);
-
+        cg::thread_block_tile<hw_warp_size>& warp) {
+        float max = to_max_float(cur_max);
         reduce::partitioned_block<rop::Max, threads_per_group>(tb, warp, max);
         Params<Type::Symmetric, numBits> params(max);
-
         return params;
     }
 };
 
-template <>
-class GroupStats<Type::Asymmetric> {
+// Asymmetric Quantization
+template <typename DataType>
+class GroupStats<Type::Asymmetric, DataType> {
 public:
-    __half2 cur_max;
-    __half2 cur_min;
+    DataType cur_max;
```

Comment:
1. In the `loco_swizzled_quant_kernel` and `loco_dequant_reduce` functions, we use the `__half` type for `dequantize::chunk` instead of `__half2`. This is because the gradients have already been quantized to 4-bit precision, and we hypothesize that using `__half` precision for dequantization is sufficient, without needing `__half2`. The goal is to reduce computation and memory usage further, though we haven't profiled to confirm specific gains.

2. Since `__half` is used for `dequantize::chunk`, we needed a version of `quantize::GroupStats` that supports the `__half` type, using like `quantize::GroupStats<quantType, __half> stats`. By defining templates, we ensure support for the `__half` type `GroupStats` without affecting any existing usage of the original `quantize::GroupStats` methods.

Reply:
`__half2` is a vector of 2 `__half`, not a higher precision of `__half`. Using `__half` will not reduce memory usage because they are the same thing.

Reply:
@hwchen2017 U mean we shall change back the to `__half2 ` for the kernel `loco_dequant_reduce` and `loco_swizzled_quant_kernel`?

Reply:
There are no benefits to using `__half` instead of `__half2`. It's fine to leave it as is if everything is working correctly. If you don't plan to change it to `__half2`, please restore the deleted comments, and add a link to your paper in the appropriate places.